src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -0,0 +1,354 @@
+steps fs bfs resolveMupdate Merge {..} = \numSteps -> do
+    curKey <- readIORef mergeCurrentKey
+    curEntry <- readIORef mergeCurrentEntry
+    go curKey curEntry numSteps
+  where
+    go curKey curEntry n
+      | n <= 0 = do
+          writeIORef mergeCurrentKey curKey
+          writeIORef mergeCurrentEntry curEntry
+          return MergeInProgress
+    go curKey curEntry !n =
+      getFromReaders fs bfs mergeReaders >>= \case
+        Just (newKey, newEntry) | newKey == curKey -> do
+          -- still same key: don't write yet, resolve!
+          let resolvedEntry = resolveEntries resolveMupdate curEntry newEntry
+          go curKey resolvedEntry (n-1)
+        Just (newKey, newEntry) -> do
+          -- new key: write old entry!
+          writeEntry fs bfs mergeWriter curKey curEntry
+          go newKey newEntry (n-1)
+        Nothing -> do
+          -- finished reading: write last entry!
+          writeEntry fs bfs mergeWriter curKey curEntry
+          run <- Run.fromMutable fs (Run.RefCount 1) mergeWriter
+          -- All Readers have been drained, the writer finalised.
+          -- No further cleanup required.
+          return (MergeComplete run)
+
+resolveEntries :: MergeValues -> PrefixEntry h -> PrefixEntry h -> PrefixEntry h
+resolveEntries mrg e1 e2 =
+    assert (prefixEntryInvariant e1) $
+    assert (prefixEntryInvariant e2) $
+    case e1 of
+      FullEntry (Mupdate v1) ->
+        -- for monoidal resolution, we need the full second value to be present
+        FullEntry (combine mrg (Mupdate v1) (toFullEntry e2))
+      _ ->
+        -- otherwise, just take the first one (OverflowEntry can't be Mupdate!)
+        e1
+
+-- | This function should be called when discarding a 'Merge' before it
+-- was done (i.e. returned 'MergeComplete'). This removes the incomplete files
+-- created for the new run so far and avoids leaking file handles.
+--
+-- Once it has been called, do not use the 'Merge' any more!
+close ::
+     HasFS IO h
+  -> Merge (FS.Handle h)
+  -> IO ()
+close fs Merge {..} = do
+    MRun.close fs mergeWriter
+    closeReaders fs mergeReaders
+
+{-------------------------------------------------------------------------------
+  Read
+-------------------------------------------------------------------------------}
+
+-- | Abstraction for the collection of 'Reader', yielding elements in order.
+-- More precisely, that mean first ordered by their key, then by the input
+-- run they came from. This is important for resolving multiple entries with the
+-- same key into one.
+--
+-- Construct with 'newReaders', then keep calling 'getFromReaders'.
+-- If aborting early, remember to call 'closeReaders'!
+data Readers fhandle = Readers {
+      readersHeap     :: !(Heap.MutableHeap RealWorld (ReadCtx fhandle))
+      -- | Since there is always one reader outside of the heap, we need to
+      -- store it separately.
+    , readersLastRead :: !(IORef (Reader fhandle, ReaderNumber))
```

Comment:
Why not put it back into the heap immediately?

Reply:
Because we then immediately get out another element. It's just how the heap API works.

Reply:
Not sure how large the change would be, but you could put it back in the heap immediately and keep track of the last extracted value here instead. Just an idea, if it makes things simpler

Reply:
I would still also need to keep track of the last extracted reader, since I need it for the next call to `replace`. There's no way around it, one of the readers needs to be stored next to the heap.